python_modules/dagster-test/dagster_test/toys/user_computed_data_versions/external_system.py

```diff
@@ -0,0 +1,155 @@
+        """Recompute a value if its provenance is stale.
+
+        Receives asset provenance info from Dagster representing the last materialization on record.
+        The provenance is compared to the current state of code/data in the system to determine
+        whether something has changed and the asset should be recomputed.
+
+        Args:
+            key (str):
+                Identifier for the asset to materialize.
+            provenance_json (Optional[str]):
+                JSON-serialized provenance info for the last materialization of the asset. `None` if
+                there is no materialization on record for the asset.
+
+        Returns (Tuple[str, bool]):
+            A tuple of the data version for the asset and a boolean flag indicating whether the data
+            version corresponds to a memoized value (true) or a freshly computed value (false).
+        """
+        provenance = json.loads(provenance_json) if provenance_json else None
+        if (
+            not self._db.has(key)
+            or provenance is None
+            or self._is_provenance_stale(key, provenance)
+        ):
+            inputs = {dep: self._db.get(dep).value for dep in _DEPENDENCIES[key]}
+            value = _compute_value(key, inputs)
+            data_version = _get_hash(value)
+            record = _DatabaseRecord(value, data_version, self.get_code_version(key))
+            self._db.set(key, record)
+            is_memoized = False
+        else:
+            record = self._db.get(key)
+            is_memoized = True
+        return record.data_version, is_memoized
+
+    def _is_provenance_stale(self, key: str, provenance: Mapping[str, Any]) -> bool:
+        record = self._db.get(key)
+        # did code change?
+        if provenance["code_version"] != record.code_version:
+            return True
+        # was a dependency added or removed?
+        if set(provenance["input_data_versions"].keys()) != _DEPENDENCIES[key]:
+            return True
+        # did the version of a dependency change?
+        for dep_key, version in provenance["input_data_versions"].items():
+            if not self._db.has(dep_key):
+                return True
+            elif self._db.get(dep_key).data_version != version:
+                return True
+        return False
+
+
+@dataclass
+class _DatabaseRecord:
+    value: Any
+    data_version: str
+    code_version: Optional[str]
+
+
+class _Database:
+    root_path: Optional[str] = None
+    _initialized: bool = False
+
+    def ensure_exists(self) -> None:
+        if not self._initialized:
+            assert self.root_path is not None
+            if not os.path.exists(self.root_path):
+                os.mkdir(self.root_path)
+            for k, v in _SOURCE_ASSETS.items():
+                with open(self.asset_path(k), "w") as fd:  # source asset
+                    record = _DatabaseRecord(v, _get_hash(v), None)
+                    fd.write(yaml.dump(asdict(record)))
```

Comment:
doesn't matter that much, but seems kind of goofy to use yaml as a serialization format. json more natural